lib/fileio/translators/translatorMayaReference.cpp

```diff
@@ -0,0 +1,489 @@
+  // handle this number of workspaces efficiently, we might consider setting
+  // -mergeNamespacesOnClash to true.)
+  //
+    MStringArray createdNodes;
+    MString referenceCommand = MString("file"
+        " -reference"
+        " -returnNewNodes"
+        " -deferReference true"
+        " -mergeNamespacesOnClash false"
+        " -ignoreVersion"
+        " -options \"v=0;\""
+        " -namespace \"") + rigNamespaceM +
+        "\" \"" + mayaReferencePath + "\"";
+
+    TF_DEBUG(PXRUSDMAYA_TRANSLATORS).Msg("MayaReferenceLogic::LoadMayaReference prim=%s execute \"%s\"\n",
+        prim.GetPath().GetText(),
+        referenceCommand.asChar());
+    status = MGlobal::executeCommand(referenceCommand, createdNodes);
+    CHECK_MSTATUS_AND_RETURN_IT(status);
+
+    if (createdNodes.length() != 1)
+    {
+        MGlobal::displayError(MString("Expected to exactly 1 node result from reference command: ") + referenceCommand);
+        return MS::kFailure;
+    }
+
+    // Retrieve created reference node
+    MObject referenceObject;
+    MString refNode = createdNodes[0];
+    MSelectionList selectionList;
+    selectionList.add(refNode);
+    selectionList.getDependNode(0, referenceObject);
+
+    // Connect prim transform's message to reference's `associatedNode` attribute,
+    // so that the referenced nodes end up under the prim transform when loaded.
+    MFnReference refDependNode(referenceObject);
+    connectReferenceAssociatedNode(parentDag, refDependNode);
+
+  // Rename the reference node.  We want a unique name so that multiple copies
+  // of a given prim can each have their own reference edits. We make a name
+  // from the full path to the prim for which the reference is being created
+  // (and append "_RN" to the end, to indicate it's a reference node, just
+  // because).
+  //
+  MString uniqueRefNodeName = refNameFromPath(parentDag) + "_RN";
+  refDependNode.setName(uniqueRefNodeName);
+
+    // Now load the reference to properly trigger the kAfterReferenceLoad callback
+    MFileIO::loadReferenceByNode(referenceObject, &status);
+    CHECK_MSTATUS_AND_RETURN_IT(status);
+    {
+        // To avoid the error that USD complains about editing to same layer simultaneously from different threads,
+        // we record it as custom data instead of creating an attribute.
+        MString refDependNodeName = refDependNode.name();
+        VtValue value(std::string(refDependNodeName.asChar(), refDependNodeName.length()));
+        prim.SetCustomDataByKey(maya_associatedReferenceNode, value);
+    }
+
+    return MS::kSuccess;
+}
+
+MStatus 
+UsdMayaTranslatorMayaReference::UnloadMayaReference(const MObject& parent){
+    TF_DEBUG(PXRUSDMAYA_TRANSLATORS).Msg("MayaReferenceLogic::UnloadMayaReference\n");
+    MStatus status;
+    MFnDependencyNode fnParent(parent, &status);
+    if (status)
+    {
+        MPlug messagePlug = fnParent.findPlug("message", true, &status);
```

Comment:
Do we need to look this attribute up by name? Maybe use an MNodeClass upfront to grab the MObject for the attribute, and initialise the MPlug directly (thereby avoiding the need for fnParent completely)